### ISSUE TITLE
with and rewrite do not abstract levels in sorts of types.

### ISSUE THREAD

**Issue by effectfully** [agda/agda#2341]
```agda
open import Level
open import Function using (id)
open import Relation.Binary.PropositionalEquality using (_≡_)
open import Relation.Binary.PropositionalEquality.TrustMe

Coerce : ∀ {β α} -> Set α -> Set β
Coerce {β} {α} rewrite trustMe {x = α} {β} = id
```

The error is

    Set (suc .w) != Set (suc α ⊔ suc .w)
    when checking that the expression id has type Set .w → Set .w

This:

```agda
Coerce' : ∀ {β α} -> Set α -> Set β
Coerce' {β} {α} rewrite trustMe {x = α} {β} = λ x -> x
```

type checks fine.

This is regression: `Coerce` was type checked successfully by a previous version of Agda, but I can't say by which one.

**Comment by nad** [comment]
The following code is accepted by Agda 2.4.2.5:
```agda
open import Agda.Primitive

data _≡_ {a} {A : Set a} (x : A) : A → Set a where
  refl : x ≡ x

{-# BUILTIN EQUALITY _≡_ #-}
{-# BUILTIN REFL refl #-}

primitive primTrustMe : ∀ {a} {A : Set a} {x y : A} → x ≡ y

id : ∀ {a} {A : Set a} → A → A
id x = x

Coerce : ∀ {β α} → Set α → Set β
Coerce {β} {α} rewrite primTrustMe {x = α} {β} = id
```

**Comment by asr** [comment]
And @nad example is rejected by Agda 2.5.1 with the error:

```
/some-path/Bug.agda:16,50-52
Set (lsuc β) != Set (lsuc α ⊔ lsuc β)
when checking that the expression id has type Set β → Set β
```

**Comment by andreasabel** [comment]
This seems to be independent of primTrustMe, it is probably a bug introduced in the change of rewrite:
```agda
open import Agda.Primitive

data _≡_ {a} {A : Set a} (x : A) : A → Set a where
  refl : x ≡ x

{-# BUILTIN EQUALITY _≡_ #-}
{-# BUILTIN REFL refl #-}

postulate
  equal : ∀ {a} {A : Set a} (x y : A) → x ≡ y

id : ∀ {a} {A : Set a} → A → A
id x = x

Works : ∀ β α → Set α → Set β
Works β α with equal α β
Works β .β | refl = id

Coerce : ∀ β α → Set α → Set β
Coerce β α rewrite equal α β = id
```

**Comment by andreasabel** [comment]
The sorts in the Pi-types in the generated with function are wrong:
```
type before rewrite:
Set α →{1+β⊔α} Set β
added with function .Issue2341.rewrite-42 of type
  (β α w : Level) → w ≡ β → Set w → Set β
type after rewrite:
  (β α : Level) → (w : Level) →{1+β⊔α} w ≡ β →{1+β⊔α} Set w →{1+β⊔α} Set β
```

**Comment by andreasabel** [comment]
Here is a simpler regression wrt. Agda-2.4.2.5:
```agda
open import Agda.Primitive

test : ∀ a → Set (lsuc a)
test a with a
... | w = Set w
```
fails with
```
w != a of type Level
when checking that the expression Set w has type Set (lsuc w)
```

**Comment by andreasabel** [comment]
Blame on https://github.com/agda/agda/commit/4a849c4404ef78a57bfebf75516b82e8ed8df59d for
```haskell
-- Doesn't abstract in the sort.
abstractType :: Type -> Term -> Type -> TCM Type
abstractType a v (El s b) = El (raise 1 s) <$> abstractTerm a v (sort s) b
```
Abstracting in the sort seems necessary.

**Comment by andreasabel** [comment]
@asr: Maybe this regression fix can be cherry-picked to the 2.5.2-release branch?!

**Comment by asr** [comment]
> Maybe this regression fix can be cherry-picked to the 2.5.2-release branch?!

Done in 72335c0070281a6a1b818ab11a68205dba6596d7.

**Comment by andreasabel** [comment]
Thanks!
